XC-7: Sales order calculation status

sethobey opened the pull request:
### Context
<!-- Why is this change necessary? Write one or two sentences to explain what's going on. -->
Refactors TaxJar Sales Order view additional info.

### Description
<!-- What does this PR change? If it's a bug, describe the fix. If it's a feature, post screenshots or a video. -->
- Sales Order TaxJar Info's section "Sync Status" displays one of three potential messages depending on order status and sync status
- Sales Order TaxJar Info's section "Calculation Status" displays one of three potential messages depending on order extension metadata's `tax_result` value.
  - Order and Creditmemo calculation status will continue to be displayed in sales_order_view and sales_order_creditimemo_view as it has been historically.
- To align with recent changes to similar feature implementation in WooCommerce plugin, PR does not include "Advanced" section. Order Metadata does not store JSON request/response body.

### Performance
<!-- How does this PR impact the area that's being changed? Prove it out. This can be an informal benchmark, EXPLAIN ANALYZE output, etc. -->
Performance impact should be negligible, but feature adds an additional DB write operation for the order extension data.

### Testing
<!-- How do we test this PR? **This section is critical.** Some ideas:
- Provide clear steps to reproduce w/ test data
- Show us how you tested the PR
- Call out specific areas of concern
-->

#### Tax not calculated
1. Checkout as Customer with TJ extension disabled
2. Enable TaxJar extension
3. Navigate to sales order view as Admin
4. Observe calculation status

![image](https://user-images.githubusercontent.com/47947793/150628787-6c53b859-15fc-4341-b4e3-9246317ce64a.png)

#### Tax calculate success
1. Checkout as Customer with TJ extension enabled for checkout
2. Navigate to sales order view as Admin
3. Click "TaxJar Information" tab
4. Observe calculation status

![image](https://user-images.githubusercontent.com/47947793/150628814-3eaba213-9b87-4aa9-96a0-868742cfd0d3.png)

#### Tax calculate error
1. Simulate API failure. e.g. modify code in `SmartCalcs.php` to throw exception on API call
2. Checkout as Customer with TJ extension enabled for checkout
3. Navigate to sales order view as Admin
4. Click "TaxJar Information" tab
5. Observe calculation status

![image](https://user-images.githubusercontent.com/47947793/150628839-79e62385-f190-44a2-9db3-a782ca58d516.png)

#### Sync status feature disabled
1. Navigate to sales order view as Admin with TJ extension transaction sync disabled
2. Click "TaxJar Information" tab
3. Observe sync status

![image](https://user-images.githubusercontent.com/47947793/150190554-9dc50f77-94a9-44df-8a57-af915590bf72.png)

#### Sync status awaiting
1. Checkout as Customer with TJ extension transaction sync enabled
2. Navigate to sales order view as Admin
3. Click "TaxJar Information" tab
4. Observe sync status ![image](https://user-images.githubusercontent.com/47947793/150189368-356fb6cc-e2a3-4687-8de6-5d96232c5379.png)
5. Invoice order ![image](https://user-images.githubusercontent.com/47947793/150189562-d0855904-ce4a-4668-bb3b-ade32757484a.png)

#### Sync status success
1. Using previous order, ship order to move status to complete
2. Click "TaxJar Information" tab
3. Observe sync status

![image](https://user-images.githubusercontent.com/47947793/150628874-6184aa9d-f765-4120-99a1-34265bf8054b.png)

#### Versions
<!-- What version(s) did you test this change on? -->
- [X] Magento 2.4
- [ ] Magento 2.3
<!-- What edition(s) of Magento did you test this change on? -->
- [X] Magento Open Source (CE)
- [ ] Magento Commerce (EE)
- [ ] Magento B2B
- [ ] Magento Cloud
<!-- What version of PHP did you test this change on? -->
- [X] PHP 7.x

sethobey-stripe committed:
feat: TaxJar sales order metadata calculation UI

sethobey-stripe committed:
refactor: Update sales order UI

sethobey-stripe committed:
refactor: Store smartcalcs calculation error in order metadata

dallendalton reviewed (changes requested):
Everything I tested looked good. However, I think we might want to make some improvements. Specifically in Model/Smartcalcs.php in the getTaxForOrder method. Everywhere where this method short circuits and returns (due to no nexus, etc.) we aren't actually performing a tax calculation request and are just falling back to Magento rates. When an order is created from one of these quotes, the calculation status message is "Tax was calculated in realtime through the TaxJar API." which is not really true for the order. 

Most often when customers think an order should have had tax calculated and it didn't it wasn't due to an unknown exception (which this PR currently handles) but rather due to our validation that short circuits the method. 

I think we need to surface the reason why we didn't calculate tax and that we reverted to Magento default tax calculations when this occurs in the metadata.

sethobey-stripe committed:
refactor: Update Smartcalcs validation message logging

sethobey commented:
> Everything I tested looked good. However, I think we might want to make some improvements. Specifically in Model/Smartcalcs.php in the getTaxForOrder method. Everywhere where this method short circuits and returns (due to no nexus, etc.) we aren't actually performing a tax calculation request and are just falling back to Magento rates. When an order is created from one of these quotes, the calculation status message is "Tax was calculated in realtime through the TaxJar API." which is not really true for the order.
> 
> Most often when customers think an order should have had tax calculated and it didn't it wasn't due to an unknown exception (which this PR currently handles) but rather due to our validation that short circuits the method.
> 
> I think we need to surface the reason why we didn't calculate tax and that we reverted to Magento default tax calculations when this occurs in the metadata.

Hey @dallendalton,
Good call out there. I've updated Smartcalcs request validation and now we should show a message like so:

![image](https://user-images.githubusercontent.com/47947793/151391438-64765646-91f0-4a65-bd57-37c7d1bfd889.png)

jordan-k-johnson reviewed (approved):
LGTM

dallendalton commented on Model/Smartcalcs.php:
```diff
@@ -543,21 +576,24 @@ private function _getLineItems(
-            ];
-        }
+        $nexusCollection = $this->nexusFactory->create()->getCollection()->addStoreFilter($storeId);
+        return array_map([$this, '_getNexusAddress'], $nexusCollection);
```
I'm getting an error here:
[2022-01-31 17:31:28] main.CRITICAL: Report ID: webapi-61f81cf011158; Message: Warning: array_map(): Expected parameter 2 to be an array, object given in /var/www/html/app/code/Taxjar/SalesTax/Model/Smartcalcs.php on line 580 {"exception":"[object] (Exception(code: 0): Report ID: webapi-61f81cf011158; Message: Warning: array_map(): Expected parameter 2 to be an array, object given in /var/www/html/app/code/Taxjar/SalesTax/Model/Smartcalcs.php on line 580 at /var/www/html/vendor/magento/framework/Webapi/ErrorProcessor.php:208, Exception(code: 0): Warning: array_map(): Expected parameter 2 to be an array, object given in /var/www/html/app/code/Taxjar/SalesTax/Model/Smartcalcs.php on line 580 at /var/www/html/vendor/magento/framework/App/ErrorHandler.php:61)"} []

sethobey replied:
🤦 I believe this was due to a bad "soft" reset.. I've abstracted these nexus-related operations into a separate Helper class and added test coverage in `\Taxjar\SalesTax\Test\Unit\Helper\NexusTest.php` to prevent regression.

sethobey-stripe committed:
fix: typo

sethobey-stripe committed:
refactor: Smartcalcs nexus helper

sethobey-stripe committed:
refactor: Remove tj metadata created_at

sethobey commented:
Also worth noting, since opening this PR, I've discovered some differences in how `Magento_Sales` implements extension attributes vs. other core-modules' extensible interfaces.

The addition of a `created_at` column to our metadata table can lead to an ambiguous `where`-clause in some situations. While this could be mitigated in other ways, such as `di.xml` configuration or use of a plugin, it seems more practical to remove the offending column since it is not necessary rather than implement additional code.

dallendalton commented on Helper/Nexus.php:
```diff
@@ -0,0 +1,109 @@
+    {
+        /** @var array|\Taxjar\SalesTax\Api\Data\Tax\NexusInterface[] $nexusArray */
+        $nexusArray = $this->getNexusCollection($storeId)->getItems();
+        return array_map(fn ($nexus) => $this->getNexusData($nexus), $nexusArray);
```
When making a request to v2/taxes I am getting a 406 response with this body:

{"status":"406","error":"Not Acceptable","detail":"nexus_addresses is invalid"}

The nexus_addresses parameter for that request contains this data:

{"42":{"id":"42","country":"US","zip":"84651","state":"UT","city":"Payson","street":"123 Main St"}}

I believe that having it start at index 42 may be the problem (haven't tested that though) and that it comes from the output of this method.

sethobey replied:
Fixed and updated unit tests to account for this

sethobey-stripe committed:
fix: Nexus helper collection array indices
